stock_packaging_calculator/models/product.py

```diff
@@ -1,51 +1,76 @@
 # Copyright 2020 Camptocamp SA
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
 
+from collections import namedtuple
+
 from odoo import models
 from odoo.tools import float_compare
 
+# Unify records as we mix up w/ UoM
+Packaging = namedtuple("Packaging", "id name qty")
+
 
 class Product(models.Model):
     _inherit = "product.product"
 
-    def product_qty_by_packaging(self, prod_qty, min_unit=None):
+    def product_qty_by_packaging(self, prod_qty, with_contained=False):
         """Calculate quantity by packaging.
 
+        The minimal quantity is always represented by the UoM of the product.
+
         Limitation: fractional quantities are lost.
 
         :prod_qty: total qty to satisfy.
-        :min_unit: minimal unit of measure as a tuple (qty, name).
-                   Default: to UoM unit.
-        :returns: list of tuple in the form [(qty_per_package, package_name)]
+        :with_contained: include calculation of contained packagings.
+
+            eg: 1 pallet contains 4 big boxes and 6 little boxes.
 
+        :returns: list of dict in the form
+
+            [{id: 1, qty: qty_per_package, name: package_name}]
+
+            If `with_contained` is passed, each element will include
+            the quantity of smaller packaging, like:
+
+            {contained: [{id: 1, qty: 4, name: "Big box"}]}
         """
-        packagings = [(x.qty, x.name) for x in self.packaging_ids]
-        if min_unit is None:
-            # You can pass `False` to skip it.
-            single_unit = self.uom_id
-            min_unit = (single_unit.factor, single_unit.name)
-        if min_unit:
-            packagings.append(min_unit)
+        packagings = [Packaging(x.id, x.name, x.qty) for x in self.packaging_ids]
+        # Add minimal unit
+        packagings.append(
+            # NOTE: the ID here could clash w/ one of the packaging's.
+            # If you create a mapping based on IDs, keep this in mind.
+            Packaging(self.uom_id.id, self.uom_id.name, self.uom_id.factor)
```

Comment:
should we add a "model" key in the dictionary so we know to what model the id refers?

Reply:
I thought about a flag instead but then it's quite easy to get the unit elem: it's always the last one.